File: apps/api/app/detectors.py
from __future__ import annotations

import hashlib
import math
import re
import statistics
from dataclasses import dataclass
from pathlib import Path
from typing import Any


DISCLAIMER = "TruthLens MVP provides probabilistic demo signals, not legal or forensic proof."


@dataclass(frozen=True)
class DetectorOutput:
    modality: str
    layer_scores: dict[str, float]
    watermark_signals: dict[str, Any]
    layer_breakdown: list[dict[str, Any]]
    detected_sources: list[str]
    artifacts: dict[str, Any]
    explanation_parts: list[str]


def clamp(value: float, low: float = 0.0, high: float = 1.0) -> float:
    return max(low, min(high, value))


def round_score(value: float) -> float:
    return round(clamp(value), 3)


def stable_unit(seed: bytes | str, salt: str) -> float:
    if isinstance(seed, str):
        seed = seed.encode("utf-8", errors="ignore")
    digest = hashlib.sha256(seed[:16384] + salt.encode("utf-8")).digest()
    return int.from_bytes(digest[:8], "big") / float(2**64 - 1)


def byte_entropy(data: bytes) -> float:
    if not data:
        return 0.0
    sample = data[: min(len(data), 262_144)]
    counts = [0] * 256
    for byte in sample:
        counts[byte] += 1
    entropy = 0.0
    total = len(sample)
    for count in counts:
        if count:
            probability = count / total
            entropy -= probability * math.log2(probability)
    return entropy / 8.0


def likely_sources(*scores: tuple[str, float]) -> list[str]:
    selected = [name for name, score in scores if score >= 0.58]
    return selected[:3] or ["No specific generator identified"]


class ImageDemoDetector:
    allowed_extensions = {".jpg", ".jpeg", ".png", ".webp", ".heic"}

    def analyze(self, data: bytes, filename: str) -> DetectorOutput:
        name = filename.lower()
        entropy = byte_entropy(data)
        size_factor = clamp(len(data) / (6 * 1024 * 1024))
        filename_ai_hint = 1.0 if re.search(r"(midjourney|stable|diffusion|dall|flux|firefly|gemini|generated|ai)", name) else 0.0
        has_exif = b"Exif" in data[:65536] or b"xmp" in data[:65536].lower()
        has_c2pa = b"c2pa" in data.lower()[:262144] or b"content credentials" in data.lower()[:262144]
        jpeg_blocks = data.count(b"\xff\xd8") + data.count(b"\xff\xdb") + data.count(b"\xff\xc4")

        neural = 0.34 + stable_unit(data, "image-neural") * 0.45 + filename_ai_hint * 0.33
        frequency = 0.22 + abs(entropy - 0.72) * 0.9 + stable_unit(data, "image-frequency") * 0.18 + filename_ai_hint * 0.08
        forensic = 0.18 + (0.16 if not has_exif else -0.08) + size_factor * 0.08 + stable_unit(data, "image-forensic") * 0.34 + filename_ai_hint * 0.08
        provenance = 0.08 + filename_ai_hint * 0.78 + (0.08 if has_c2pa else 0.0)
        watermark = 0.1 + (0.36 if b"synthid" in data.lower()[:262144] else 0.0) + filename_ai_hint * 0.14

        layer_scores = {
            "watermark": round_score(watermark),
            "forensic": round_score(forensic),
            "neural_classifier": round_score(neural),
            "frequency_domain": round_score(frequency),
            "provenance": round_score(provenance),
        }

        heatmap = [
            {
                "x": x,
                "y": y,
                "score": round_score(0.22 + stable_unit(data, f"heatmap-{x}-{y}") * 0.68),
            }
            for y in range(6)
            for x in range(8)
        ]

        if filename_ai_hint or has_c2pa:
            detected_sources = likely_sources(
                ("Midjourney-style image", layer_scores["neural_classifier"]),
                ("Stable Diffusion / Flux-style image", layer_scores["frequency_domain"]),
                ("C2PA-attributed AI image", 0.72 if has_c2pa and filename_ai_hint else 0.0),
            )
        else:
            detected_sources = ["No specific generator identified"]

        explanation_parts = [
            "Image analysis used deterministic MVP heuristics for metadata, byte distribution, filename hints, and pseudo-neural scoring.",
            "No real SynthID, C2PA verification, or AIDE model inference is integrated in this build.",
        ]
        if not has_exif:
            explanation_parts.append("No EXIF/XMP camera metadata was found in the inspected byte range, which is treated as weak synthetic-media evidence.")
        if filename_ai_hint:
            explanation_parts.append("The filename contains AI-generator wording, which increases provenance suspicion in this demo mode.")

        return DetectorOutput(
            modality="IMAGE",
            layer_scores=layer_scores,
            watermark_signals={
                "synthid": {"detected": False, "score": layer_scores["watermark"], "status": "demo_not_integrated"},
                "c2pa": {"present": has_c2pa, "valid": None, "status": "demo_metadata_scan"},
            },
            layer_breakdown=[
                {"name": "Watermark", "score": layer_scores["watermark"], "explanation": "Demo scan for watermark-like byte markers only."},
                {"name": "Forensic", "score": layer_scores["forensic"], "explanation": "Metadata presence, size, and byte-distribution heuristics."},
                {"name": "Neural Classifier", "score": layer_scores["neural_classifier"], "explanation": "Deterministic pseudo-neural score for UI/API validation."},
                {"name": "Frequency Domain", "score": layer_scores["frequency_domain"], "explanation": "Entropy-derived stand-in for FFT/DCT anomaly scoring."},
                {"name": "Provenance", "score": layer_scores["provenance"], "explanation": "Filename and demo C2PA marker checks."},
            ],
            detected_sources=detected_sources,
            artifacts={
                "heatmap": heatmap,
                "metadata": {"has_exif_or_xmp": has_exif, "jpeg_marker_count": jpeg_blocks, "byte_entropy": round(entropy, 3)},
            },
            explanation_parts=explanation_parts,
        )


class VideoDemoDetector:
    allowed_extensions = {".mp4", ".mov", ".avi", ".webm"}

    def analyze(self, data: bytes, filename: str) -> DetectorOutput:
        name = filename.lower()
        entropy = byte_entropy(data)
        size_mb = len(data) / (1024 * 1024)
        filename_ai_hint = 1.0 if re.search(r"(sora|veo|runway|kling|deepfake|generated|ai)", name) else 0.0
        has_mp4_metadata = any(marker in data[:262144].lower() for marker in [b"moov", b"mvhd", b"udta"])

        temporal = 0.26 + stable_unit(data, "video-temporal") * 0.45 + filename_ai_hint * 0.18
        face_mesh = 0.24 + stable_unit(data, "video-face") * 0.42 + filename_ai_hint * 0.14
        neural = 0.3 + stable_unit(data, "video-neural") * 0.42 + clamp(size_mb / 200) * 0.06
        provenance = 0.1 + filename_ai_hint * 0.45 + (0.07 if not has_mp4_metadata else 0.0)
        watermark = 0.08 + (0.34 if b"synthid" in data.lower()[:262144] or b"videoseal" in data.lower()[:262144] else 0.0)

        layer_scores = {
            "watermark": round_score(watermark),
            "temporal_forensics": round_score(temporal),
            "face_mesh": round_score(face_mesh),
            "neural_classifier": round_score(neural),
            "provenance": round_score(provenance),
        }

        segments = []
        for index in range(8):
            start = index * 5
            score = round_score(0.2 + stable_unit(data, f"video-segment-{index}") * 0.68)
            segments.append({"start_seconds": start, "end_seconds": start + 5, "score": score})

        detected_sources = likely_sources(
            ("Runway/Kling-style generated video", layer_scores["temporal_forensics"]),
            ("Sora/Veo-style generated video", layer_scores["neural_classifier"]),
            ("Face-swap/deepfake-style edit", layer_scores["face_mesh"]),
        )

        return DetectorOutput(
            modality="VIDEO",
            layer_scores=layer_scores,
            watermark_signals={
                "synthid_video": {"detected": False, "score": layer_scores["watermark"], "status": "demo_not_integrated"},
                "video_seal": {"detected": False, "score": layer_scores["watermark"], "status": "demo_not_integrated"},
                "c2pa": {"present": False, "valid": None, "status": "demo_not_integrated"},
            },
            layer_breakdown=[
                {"name": "Watermark", "score": layer_scores["watermark"], "explanation": "Demo marker scan for SynthID/VideoSeal-like strings."},
                {"name": "Temporal Forensics", "score": layer_scores["temporal_forensics"], "explanation": "Deterministic stand-in for optical-flow and coherence analysis."},
                {"name": "Face Mesh", "score": layer_scores["face_mesh"], "explanation": "Pseudo-score for face geometry and lip-sync inconsistency checks."},
                {"name": "Neural Classifier", "score": layer_scores["neural_classifier"], "explanation": "Deterministic pseudo-neural score for video generator coverage."},
                {"name": "Provenance", "score": layer_scores["provenance"], "explanation": "Container marker and filename provenance heuristics."},
            ],
            detected_sources=detected_sources,
            artifacts={
                "timeline": segments,
                "metadata": {"size_mb": round(size_mb, 2), "byte_entropy": round(entropy, 3), "has_mp4_metadata": has_mp4_metadata},
            },
            explanation_parts=[
                "Video analysis used deterministic MVP heuristics for byte patterns, container markers, filename hints, and timeline scoring.",
                "Real frame extraction, MediaPipe face mesh, optical flow, VideoSeal, and deepfake model inference are deferred.",
            ],
        )


class TextHeuristicDetector:
    allowed_extensions = {".txt", ".md", ".docx", ".pdf"}
    ai_phrases = {
        "in conclusion",
        "it is important to note",
        "moreover",
        "furthermore",
        "delve",
        "realm",
        "underscores",
        "plays a crucial role",
        "comprehensive",
        "seamless",
        "as an ai",
    }

    def analyze(self, text: str, filename: str | None = None) -> DetectorOutput:
        normalized = re.sub(r"\s+", " ", text.strip())
        sentences = [sentence.strip() for sentence in re.split(r"(?<=[.!?])\s+", normalized) if sentence.strip()]
        words = re.findall(r"[A-Za-z][A-Za-z'-]*", normalized.lower())
        word_count = len(words)
        unique_ratio = len(set(words)) / max(word_count, 1)
        sentence_lengths = [len(re.findall(r"[A-Za-z][A-Za-z'-]*", sentence)) for sentence in sentences] or [word_count]
        avg_sentence = statistics.mean(sentence_lengths)
        burstiness = statistics.pstdev(sentence_lengths) / max(avg_sentence, 1)
        phrase_hits = sum(normalized.lower().count(phrase) for phrase in self.ai_phrases)
        phrase_density = phrase_hits / max(len(sentences), 1)
        punctuation_uniformity = 1.0 - clamp(len(set(sentence[-1] for sentence in sentences if sentence)) / 4)

        burstiness_reliability = clamp((len(sentences) - 2) / 3)
        low_burstiness_signal = clamp(1.0 - burstiness) * burstiness_reliability
        polished_length_signal = clamp((avg_sentence - 12) / 18)
        lexical_signal = clamp((0.62 - unique_ratio) * 2.2)
        phrase_signal = clamp(phrase_density / 0.6)
        first_person_count = sum(1 for word in words if word in {"i", "me", "my", "mine", "we", "our", "ours"})
        human_context_signal = clamp((first_person_count / max(word_count, 1)) * 12 + clamp(unique_ratio - 0.72) * 1.5)

        linguistic = 0.2 + low_burstiness_signal * 0.22 + polished_length_signal * 0.18 + phrase_signal * 0.42
        if phrase_hits == 0:
            linguistic -= human_context_signal * 0.2
        neural = 0.26 + stable_unit(normalized, "text-neural") * 0.12 + linguistic * 0.58 + phrase_signal * 0.08
        if phrase_hits == 0:
            neural -= human_context_signal * 0.14
        provenance = 0.08 + (0.12 if filename and Path(filename).suffix.lower() in {".docx", ".pdf"} else 0.0)
        watermark = 0.08 + (0.32 if "synthid" in normalized.lower() else 0.0)

        layer_scores = {
            "watermark": round_score(watermark),
            "linguistic": round_score(linguistic),
            "neural_classifier": round_score(neural),
            "provenance": round_score(provenance),
        }

        sentence_artifacts = []
        for index, sentence in enumerate(sentences[:80]):
            length = max(len(re.findall(r"[A-Za-z][A-Za-z'-]*", sentence)), 1)
            phrase_bonus = 0.2 if any(phrase in sentence.lower() for phrase in self.ai_phrases) else 0.0
            score = round_score(0.18 + clamp((length - 10) / 28) * 0.28 + phrase_bonus + stable_unit(sentence, "sentence") * 0.26)
            sentence_artifacts.append({"index": index, "text": sentence, "score": score})

        detected_sources = likely_sources(
            ("General LLM-style prose", layer_scores["linguistic"]),
            ("GPT-style assistant prose", phrase_signal),
            ("Gemini/SynthID text candidate", layer_scores["watermark"]),
        )

        explanation_parts = [
            "Text analysis used deterministic MVP heuristics for sentence rhythm, burstiness, lexical variety, phrase signatures, and pseudo-neural scoring.",
            "Real SynthID text detection and transformer classifiers are not integrated in this build.",
        ]
        if phrase_hits:
            explanation_parts.append("Repeated assistant-style transition phrases increased the AI-likelihood score.")
        if low_burstiness_signal > 0.65:
            explanation_parts.append("Sentence lengths are unusually uniform, a weak signal often associated with model-generated text.")

        return DetectorOutput(
            modality="TEXT",
            layer_scores=layer_scores,
            watermark_signals={
                "synthid_text": {"detected": "synthid" in normalized.lower(), "score": layer_scores["watermark"], "status": "demo_marker_scan"},
            },
            layer_breakdown=[
                {"name": "Watermark", "score": layer_scores["watermark"], "explanation": "Demo marker scan for SynthID-like wording only."},
                {"name": "Linguistic", "score": layer_scores["linguistic"], "explanation": "Burstiness, sentence rhythm, lexical variety, and phrase-density heuristics."},
                {"name": "Neural Classifier", "score": layer_scores["neural_classifier"], "explanation": "Deterministic pseudo-neural score for API and UI validation."},
                {"name": "Provenance", "score": layer_scores["provenance"], "explanation": "Document-source context only; no text provenance standard is verified."},
            ],
            detected_sources=detected_sources,
            artifacts={
                "sentences": sentence_artifacts,
                "metrics": {
                    "word_count": word_count,
                    "sentence_count": len(sentences),
                    "lexical_diversity": round(unique_ratio, 3),
                    "burstiness": round(burstiness, 3),
                    "phrase_hits": phrase_hits,
                    "punctuation_uniformity": round(punctuation_uniformity, 3),
                    "human_context": round(human_context_signal, 3),
                },
            },
            explanation_parts=explanation_parts,
        )

File: apps/web/lib/server/detectors.ts
import { createHash, randomUUID } from "node:crypto";
import type { AnalysisResponse, HeatmapCell, LayerBreakdown, SentenceArtifact, TimelineSegment } from "../types";

const disclaimer = "TruthLens provides probabilistic AI-content signals, not legal or forensic proof.";

export const calibrationVersion = "hf-model-primary-v1";

export interface DetectorOutput {
  modality: AnalysisResponse["modality"];
  layer_scores: Record<string, number>;
  watermark_signals: Record<string, unknown>;
  layer_breakdown: LayerBreakdown[];
  detected_sources: string[];
  artifacts: AnalysisResponse["artifacts"];
  explanation_parts: string[];
}

export interface VideoFrameInput {
  data: Buffer;
  index: number;
  seconds: number;
}

const aiPhrases = [
  "in conclusion",
  "it is important to note",
  "moreover",
  "furthermore",
  "delve",
  "realm",
  "underscores",
  "plays a crucial role",
  "comprehensive",
  "seamless",
  "as an ai"
];

export function clamp(value: number, low = 0, high = 1) {
  return Math.max(low, Math.min(high, value));
}

export function roundScore(value: number) {
  return Math.round(clamp(value) * 1000) / 1000;
}

function stableUnit(seed: Buffer | string, salt: string) {
  const source = Buffer.isBuffer(seed) ? seed.subarray(0, 16384) : Buffer.from(seed, "utf8");
  const digest = createHash("sha256").update(source).update(salt).digest();
  return (digest.readUInt32BE(0) * 0x100000000 + digest.readUInt32BE(4)) / 0xffffffffffffffff;
}

function byteEntropy(data: Buffer) {
  if (!data.length) return 0;
  const sample = data.subarray(0, Math.min(data.length, 262144));
  const counts = new Array<number>(256).fill(0);
  for (let index = 0; index < sample.length; index += 1) {
    counts[sample[index]] += 1;
  }
  let entropy = 0;
  for (const count of counts) {
    if (count > 0) {
      const probability = count / sample.length;
      entropy -= probability * Math.log2(probability);
    }
  }
  return entropy / 8;
}

function likelySources(scores: Array<[string, number]>) {
  const selected = scores.filter(([, score]) => score >= 0.58).map(([name]) => name);
  return selected.slice(0, 3).length ? selected.slice(0, 3) : ["No specific generator identified"];
}

function verdictFor(score: number): AnalysisResponse["verdict"] {
  if (score >= 0.85) return "AI_GENERATED";
  if (score >= 0.62) return "LIKELY_AI";
  if (score >= 0.42) return "UNCERTAIN";
  if (score >= 0.22) return "LIKELY_HUMAN";
  return "HUMAN";
}

function sentenceWordCount(sentence: string) {
  return (sentence.match(/[A-Za-z][A-Za-z'-]*/g) ?? []).length;
}

function metadataFor(output: DetectorOutput) {
  output.artifacts.metadata ??= {};
  return output.artifacts.metadata;
}

function modelScoreFromMetadata(output: DetectorOutput) {
  const metadata = output.artifacts.metadata;
  return metadata?.hf_status === "ok" && typeof metadata.hf_score === "number" ? metadata.hf_score : null;
}

function confidenceRange(score: number, output: DetectorOutput): [number, number] {
  const values = Object.values(output.layer_scores);
  const spread = values.length ? Math.max(...values) - Math.min(...values) : 0;
  const modelPrimary = modelScoreFromMetadata(output) !== null;
  const width = modelPrimary ? clamp(0.04 + spread * 0.08, 0.04, 0.12) : clamp(0.08 + spread * 0.14, 0.08, 0.18);
  return [roundScore(score - width), roundScore(score + width)];
}

function weightedFallbackScore(output: DetectorOutput) {
  const values = Object.values(output.layer_scores);
  if (!values.length) return 0.5;
  return roundScore(values.reduce((total, value) => total + value, 0) / values.length);
}

function fuse(output: DetectorOutput, processingTimeMs: number): AnalysisResponse {
  const metadata = metadataFor(output);
  const modelScore = modelScoreFromMetadata(output);
  const confidence = modelScore ?? weightedFallbackScore(output);
  metadata.fusion_strategy = modelScore === null ? "heuristic_fallback" : "model_primary";
  metadata.calibration_version = calibrationVersion;

  return {
    request_id: randomUUID(),
    verdict: verdictFor(confidence),
    confidence,
    confidence_range: confidenceRange(confidence, output),
    modality: output.modality,
    detected_sources: output.detected_sources,
    watermark_signals: output.watermark_signals,
    layer_scores: output.layer_scores,
    layer_breakdown: output.layer_breakdown,
    artifacts: output.artifacts,
    explanation: output.explanation_parts.join(" "),
    processing_time_ms: processingTimeMs,
    disclaimer
  };
}

export function analyzeImage(data: Buffer, filename: string): DetectorOutput {
  const lower = data.subarray(0, 262144).toString("latin1").toLowerCase();
  const entropy = byteEntropy(data);
  const sizeFactor = clamp(data.length / (6 * 1024 * 1024));
  const filenameAiHint = /(midjourney|stable|diffusion|dall|flux|firefly|gemini|generated|ai)/i.test(filename);
  const hasExif = data.subarray(0, 65536).includes("Exif") || lower.includes("xmp");
  const hasC2pa = lower.includes("c2pa") || lower.includes("content credentials");
  const jpegMarkerCount = (data.toString("latin1").match(/\xff\xd8|\xff\xdb|\xff\xc4/g) ?? []).length;

  const layerScores = {
    watermark: roundScore(0.1 + (lower.includes("synthid") ? 0.36 : 0)),
    forensic: roundScore(0.18 + (hasExif ? -0.08 : 0.16) + sizeFactor * 0.08 + stableUnit(data, "image-forensic") * 0.34),
    neural_classifier: roundScore(0.5),
    frequency_domain: roundScore(0.22 + Math.abs(entropy - 0.72) * 0.9 + stableUnit(data, "image-frequency") * 0.18),
    provenance: roundScore(0.08 + (hasC2pa ? 0.08 : 0))
  };
  const heatmap: HeatmapCell[] = Array.from({ length: 48 }, (_, index) => ({
    x: index % 8,
    y: Math.floor(index / 8),
    score: roundScore(0.22 + stableUnit(data, `heatmap-${index % 8}-${Math.floor(index / 8)}`) * 0.68)
  }));

  return {
    modality: "IMAGE",
    layer_scores: layerScores,
    watermark_signals: {
      synthid: { detected: false, score: layerScores.watermark, status: "marker_scan_only" },
      c2pa: { present: hasC2pa, valid: null, status: "metadata_scan_only" }
    },
    layer_breakdown: [
      { name: "Watermark", score: layerScores.watermark, explanation: "Marker scan for visible SynthID-like strings." },
      { name: "Forensic", score: layerScores.forensic, explanation: "Supporting metadata and byte-distribution checks only." },
      { name: "Neural Classifier", score: layerScores.neural_classifier, explanation: "Awaiting Hugging Face image classifier inference." },
      { name: "Frequency Domain", score: layerScores.frequency_domain, explanation: "Supporting entropy anomaly signal only." },
      { name: "Provenance", score: layerScores.provenance, explanation: "C2PA marker context; filename hints are not used for scoring." }
    ],
    detected_sources: hasC2pa ? ["C2PA-attributed content"] : ["No specific generator identified"],
    artifacts: {
      heatmap,
      metadata: {
        source_filename: filename,
        filename_ai_hint: filenameAiHint,
        has_exif_or_xmp: hasExif,
        jpeg_marker_count: jpegMarkerCount,
        byte_entropy: Math.round(entropy * 1000) / 1000
      }
    },
    explanation_parts: [
      "Image scoring uses the model classifier as the primary signal when available.",
      "Metadata, provenance, and byte-distribution checks are supporting context and do not override model probability."
    ]
  };
}

export function analyzeText(text: string, filename?: string): DetectorOutput {
  const normalized = text.trim().replace(/\s+/g, " ");
  const sentences = normalized.split(/(?<=[.!?])\s+/).filter(Boolean);
  const words = normalized.toLowerCase().match(/[a-z][a-z'-]*/g) ?? [];
  const wordCount = words.length;
  const uniqueRatio = new Set(words).size / Math.max(wordCount, 1);
  const sentenceLengths = sentences.length ? sentences.map((sentence) => sentenceWordCount(sentence)) : [wordCount];
  const avgSentence = sentenceLengths.reduce((total, length) => total + length, 0) / sentenceLengths.length;
  const variance = sentenceLengths.reduce((total, length) => total + (length - avgSentence) ** 2, 0) / sentenceLengths.length;
  const burstiness = Math.sqrt(variance) / Math.max(avgSentence, 1);
  const lower = normalized.toLowerCase();
  const phraseHits = aiPhrases.reduce((total, phrase) => total + lower.split(phrase).length - 1, 0);
  const phraseSignal = clamp(phraseHits / Math.max(sentences.length, 1) / 0.6);
  const lowBurstinessSignal = clamp(1 - burstiness) * clamp((sentences.length - 2) / 3);
  const polishedLengthSignal = clamp((avgSentence - 12) / 18);
  const firstPersonCount = words.filter((word) => ["i", "me", "my", "mine", "we", "our", "ours"].includes(word)).length;
  const humanContextSignal = clamp((firstPersonCount / Math.max(wordCount, 1)) * 12 + clamp(uniqueRatio - 0.72) * 1.5);

  let linguistic = 0.2 + lowBurstinessSignal * 0.22 + polishedLengthSignal * 0.18 + phraseSignal * 0.42;
  if (phraseHits === 0) linguistic -= humanContextSignal * 0.2;

  const layerScores = {
    watermark: roundScore(0.08 + (lower.includes("synthid") ? 0.32 : 0)),
    linguistic: roundScore(linguistic),
    neural_classifier: roundScore(0.5),
    provenance: roundScore(0.08)
  };
  const sentenceArtifacts: SentenceArtifact[] = sentences.slice(0, 80).map((sentence, index) => {
    const length = Math.max(sentenceWordCount(sentence), 1);
    const phraseBonus = aiPhrases.some((phrase) => sentence.toLowerCase().includes(phrase)) ? 0.2 : 0;
    return { index, text: sentence, score: roundScore(0.18 + clamp((length - 10) / 28) * 0.28 + phraseBonus + stableUnit(sentence, "sentence") * 0.26) };
  });

  return {
    modality: "TEXT",
    layer_scores: layerScores,
    watermark_signals: { synthid_text: { detected: lower.includes("synthid"), score: layerScores.watermark, status: "marker_scan_only" } },
    layer_breakdown: [
      { name: "Watermark", score: layerScores.watermark, explanation: "Marker scan for SynthID-like wording." },
      { name: "Linguistic", score: layerScores.linguistic, explanation: "Supporting rhythm, lexical variety, and phrase-density signal." },
      { name: "Neural Classifier", score: layerScores.neural_classifier, explanation: "Awaiting Hugging Face text classifier inference." },
      { name: "Provenance", score: layerScores.provenance, explanation: "Document-source context; filename and file type are not used for scoring." }
    ],
    detected_sources: likelySources([
      ["General LLM-style prose", layerScores.linguistic],
      ["GPT-style assistant prose", phraseSignal],
      ["Gemini/SynthID text candidate", layerScores.watermark]
    ]),
    artifacts: {
      sentences: sentenceArtifacts,
      metrics: {
        word_count: wordCount,
        sentence_count: sentences.length,
        lexical_diversity: Math.round(uniqueRatio * 1000) / 1000,
        burstiness: Math.round(burstiness * 1000) / 1000,
        phrase_hits: phraseHits,
        human_context: Math.round(humanContextSignal * 1000) / 1000
      },
      metadata: {
        source_filename: filename ?? null
      }
    },
    explanation_parts: [
      "Text scoring uses the model classifier as the primary signal when available.",
      "Linguistic markers are supporting context and do not override model probability.",
      ...(phraseHits ? ["Assistant-style transition phrases were found as a supporting signal."] : []),
      ...(lowBurstinessSignal > 0.65 ? ["Sentence lengths are unusually uniform, a weak supporting signal often associated with model-generated text."] : [])
    ]
  };
}

export function analyzeVideo(data: Buffer, filename: string): DetectorOutput {
  const lower = data.subarray(0, 262144).toString("latin1").toLowerCase();
  const entropy = byteEntropy(data);
  const sizeMb = data.length / (1024 * 1024);
  const hasMp4Metadata = ["moov", "mvhd", "udta"].some((marker) => lower.includes(marker));
  const filenameAiHint = /(sora|veo|runway|kling|deepfake|generated|ai)/i.test(filename);
  const layerScores = {
    watermark: roundScore(0.08 + (lower.includes("synthid") || lower.includes("videoseal") ? 0.34 : 0)),
    temporal_forensics: roundScore(0.3),
    face_mesh: roundScore(0.3),
    neural_classifier: roundScore(0.5),
    provenance: roundScore(0.1 + (hasMp4Metadata ? 0 : 0.07))
  };
  return {
    modality: "VIDEO",
    layer_scores: layerScores,
    watermark_signals: {
      synthid_video: { detected: false, score: layerScores.watermark, status: "marker_scan_only" },
      video_seal: { detected: false, score: layerScores.watermark, status: "marker_scan_only" },
      c2pa: { present: false, valid: null, status: "not_verified" }
    },
    layer_breakdown: [
      { name: "Watermark", score: layerScores.watermark, explanation: "Marker scan for SynthID/VideoSeal-like strings." },
      { name: "Temporal Forensics", score: layerScores.temporal_forensics, explanation: "Frame-level model scoring is required for the deployed video path." },
      { name: "Face Mesh", score: layerScores.face_mesh, explanation: "Face-level deepfake analysis is not run in the Vercel frame-sampling path." },
      { name: "Neural Classifier", score: layerScores.neural_classifier, explanation: "Awaiting Hugging Face frame classifier inference." },
      { name: "Provenance", score: layerScores.provenance, explanation: "Container context only; filename hints are not used for scoring." }
    ],
    detected_sources: ["No specific generator identified"],
    artifacts: {
      timeline: [],
      metadata: {
        source_filename: filename,
        filename_ai_hint: filenameAiHint,
        size_mb: Math.round(sizeMb * 100) / 100,
        byte_entropy: Math.round(entropy * 1000) / 1000,
        has_mp4_metadata: hasMp4Metadata
      }
    },
    explanation_parts: [
      "Video scoring uses sampled frame classifier results as the primary signal when available.",
      "Raw video container bytes are treated as supporting metadata only."
    ]
  };
}

export function analyzeVideoFrames(frames: VideoFrameInput[], filename?: string): DetectorOutput {
  const frameEntropies = frames.map((frame) => byteEntropy(frame.data));
  const averageEntropy = frameEntropies.length ? frameEntropies.reduce((total, value) => total + value, 0) / frameEntropies.length : 0;
  const joinedSeed = Buffer.concat(frames.map((frame) => frame.data.subarray(0, 4096)));
  const filenameAiHint = filename ? /(sora|veo|runway|kling|deepfake|generated|ai)/i.test(filename) : false;
  const layerScores = {
    watermark: roundScore(0.08),
    temporal_forensics: roundScore(0.22 + Math.abs(averageEntropy - 0.72) * 0.45),
    face_mesh: roundScore(0.25 + stableUnit(joinedSeed, "video-face-support") * 0.16),
    neural_classifier: roundScore(0.5),
    provenance: roundScore(0.1)
  };
  const timeline: TimelineSegment[] = frames.map((frame, index) => ({
    start_seconds: frame.seconds,
    end_seconds: frames[index + 1]?.seconds ?? frame.seconds + 1,
    score: layerScores.neural_classifier
  }));

  return {
    modality: "VIDEO",
    layer_scores: layerScores,
    watermark_signals: {
      synthid_video: { detected: false, score: layerScores.watermark, status: "frame_marker_scan_only" },
      video_seal: { detected: false, score: layerScores.watermark, status: "frame_marker_scan_only" },
      c2pa: { present: false, valid: null, status: "not_verified" }
    },
    layer_breakdown: [
      { name: "Watermark", score: layerScores.watermark, explanation: "Frame marker scan for watermark-like strings." },
      { name: "Temporal Forensics", score: layerScores.temporal_forensics, explanation: "Supporting frame entropy consistency signal." },
      { name: "Face Mesh", score: layerScores.face_mesh, explanation: "Supporting frame-level face/deepfake placeholder signal." },
      { name: "Neural Classifier", score: layerScores.neural_classifier, explanation: "Awaiting Hugging Face frame classifier inference." },
      { name: "Provenance", score: layerScores.provenance, explanation: "Container and filename context only; not used for scoring." }
    ],
    detected_sources: ["No specific generator identified"],
    artifacts: {
      timeline,
      metadata: {
        source_filename: filename ?? null,
        filename_ai_hint: filenameAiHint,
        sampled_frame_count: frames.length,
        average_frame_entropy: Math.round(averageEntropy * 1000) / 1000
      }
    },
    explanation_parts: [
      "Video scoring uses browser-sampled frames and the image AI detector as the primary deployed signal.",
      "Full temporal deepfake modeling is not run in this Vercel build."
    ]
  };
}

export function buildResult(output: DetectorOutput, startedAt: number) {
  return fuse(output, Math.max(1, Date.now() - startedAt));
}
